[Inductor XPU GEMM] Step 1/N: Add cutlass-sycl repro.

etaf opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #160706
* #160688
* #160687
* #160686
* #160685
* #160174
* __->__ #160173

This PR is the first step toward implementing RFC #160175.
To ensure that the Intel Cutlass repository is only downloaded in XPU environments, we avoid using a Git submodule. Instead, we adopt the same approach as torch-xpu-ops: the repository is automatically cloned only when USE_XPU=1 is set.

The Intel Cutlass directory is named third_party/torch-xpu-tla (XPU Templates for Linear Algebra), and the commit ID is managed via a separate file: third_party/xpu-tla.txt.

etaf committed:
[Inductor Intel Cutlass] Step 1/N: Add Intel Cutlass repro into third_party.

[ghstack-poisoned]

Skylion007 commented on caffe2/CMakeLists.txt:
```diff
@@ -1125,6 +1125,36 @@ elseif(USE_CUDA)
 endif()
 
 if(USE_XPU)
+  function(clone_and_checkout REPO_URL DEST_DIR COMMIT_FILE)
```
Cmake has native support for fetching git now since CMake 3.14

'''
include(FetchContent)

FetchContent_Declare(
  some_lib
  GIT_REPOSITORY https://github.com/.../repo.git
  GIT_TAG        v1.2.3
  GIT_SHALLOW    TRUE
)
FetchContent_MakeAvailable(some_lib)
'''

etaf replied:
Hi, @Skylion007  @EikanWang , the `FetchContent_Declare` and `FetchContent_MakeAvailable` includes clone, checkout and add_subdirectory. For torc-xpu-ops, it need to specify the output binary directory when add_subdirectory: 
``` 
add_subdirectory(${TORCH_ROOT}/third_party/torch-xpu-ops
      ${CMAKE_BINARY_DIR}/caffe2/aten_xpu)
```
But `FetchContent_Declare` does not support it.
And for torch-xpu-tla (intel cutlass) we only want to the header files in the repo, no need to build it, so add_subdirectory is not need.

Base on that, I think current implementation is more flexible.

EikanWang replied:
@etaf , I cannot capture your point. Do you mean even `FetchContent_Populate` does not work?

etaf replied:
Hi, @EikanWang : 
  Calling FetchContent_Populate(torch-xpu-tla) is deprecated, call
  FetchContent_MakeAvailable(torch-xpu-tla-repo) instead.  Policy CMP0169 can
  be set to OLD to allow FetchContent_Populate(torch-xpu-tla-repo) to be
  called directly for now, but the ability to call it with declared details
  will be removed completely in a future version.

EikanWang replied:
Hmm... Thanks for the information @etaf . @Skylion007 any comments?

etaf committed:
Update on "[Inductor XPU GEMM] Step 1/N: Add cutlass-sycl repro."


This PR is the first step toward implementing RFC #160175.
To ensure that the Intel Cutlass repository is only downloaded in XPU environments, we avoid using a Git submodule. Instead, we adopt the same approach as torch-xpu-ops: the repository is automatically cloned only when USE_XPU=1 is set.

The Intel Cutlass directory is named third_party/torch-xpu-tla (XPU Templates for Linear Algebra), and the commit ID is managed via a separate file: third_party/xpu-tla.txt.

[ghstack-poisoned]

etaf committed:
Update on "[Inductor XPU GEMM] Step 1/N: Add cutlass-sycl repro."


This PR is the first step toward implementing RFC #160175.
To ensure that the Intel Cutlass repository is only downloaded in XPU environments, we avoid using a Git submodule. Instead, we adopt the same approach as torch-xpu-ops: the repository is automatically cloned only when USE_XPU=1 is set.

The Intel Cutlass directory is named third_party/torch-xpu-tla (XPU Templates for Linear Algebra), and the commit ID is managed via a separate file: third_party/xpu-tla.txt.

[ghstack-poisoned]

jansel reviewed (changes requested):
For the inductor backend, isn't the repo needed at *runtime* not compile time?  For CUTLASS we require the user to `git clone <cutlass>` then point inductor to the cutlass directory on their disk.

I think this should be handled similar to the triton pin to install it on CI machines.

etaf commented:
> For the inductor backend, isn't the repo needed at _runtime_ not compile time? For CUTLASS we require the user to `git clone <cutlass>` then point inductor to the cutlass directory on their disk.
> 
> I think this should be handled similar to the triton pin to install it on CI machines.

Hi @jansel , thanks for your review, you are right, let me close this PR.
